### ISSUE TITLE
Adding type hints to the AST?

### ISSUE THREAD

**Issue by MichaelChirico** [r-lib/lintr#901]
R is dynamically typed, which limits the scope of some linters / makes some linters impossible to write.

Nonetheless, there are some functions in R whose output we can type with certainty, e.g.

* `as.{raw,logical,integer,double,complex,character,list}` converters
* `lapply()` --> `list()`
* `paste()`/`paste0()`/`sprintf()`/`file.path()`/`gsub()` --> `character()`
* `grepl()`/`any()`/`all()`/`&&`/`||`/`&`/`|`/`xor()`/`%in%`/`nzchar()`/`==`/`!=`/`>=`/`<=`/`>`/`<` --> `logical()`
* `which()`/`grep(value=FALSE)` --> `integer()`
* ...

We could patch this sort of information as a "type hint" into the `utils::getParseData()` `data.frame()` (as a new column) and/or the `xml_parse_data()` XML tree (as XML attributes) to make it available to linters.

The downside is:

 1. Extra maintenance burden of these new structures
 2. (Part of 1 really, but worth emphasizing) Incrementing/scoping a list of functions to include
 3. Performance impact to building these augmented structures
 4. Always going to be incomplete as probably most nodes will just have missing data -- just the nature of a dynamically-typed language

What do the other maintainers think of the benefit/cost of this?

**Comment by AshesITR** [comment]
What would we do with these type hints?
Sounds nice but like a massive maintenance burden.
Also, functions can be overwritten (even in their namespace if you really want to), or change their return types across R versions. This means we could only heuristically infer types which might be wrong in some cases.

**Comment by jimhester** [comment]
I think type hinting is too broad a topic for lintr to take on.

You might want to get in touch with @krlmlr if you are interested in this type of type annotation. He has a working group at https://github.com/pre-processing-r that is discussing things to do with imputed and annotated types.

**Comment by MichaelChirico** [comment]
> Also, functions can be overwritten (even in their namespace if you really want to), or change their return types across R versions. This means we could only heuristically infer types which might be wrong in some cases.

Definitely true, but we can also be conservative. If someone overwrites `lapply()` not to return a list... they can deal with the consequences.

A good example is our recent `ListComparisonLinter`. It looks for `lapply(x, FUN) > y`, and other cases where comparison operators are used in conjunction with known-list functions. We manually supplied `lapply`, `map`, and `Map` as the most common likely culprits, but the linter would be more powerful if we maintained a list of such known-list functions more centrally, which could also be shared with other linters.

I think it's probably too much to bother with for now. Maybe v4.0.0 :)
